doc: add EDA and data validation sections with other minor updates (#655)
Azure-Samples/modern-data-warehouse-dataops: single_tech_samples/fabric/feature_engineering_on_fabric/README.md

    - [Data lineage](#data-lineage)
    - [Feature lineage](#feature-lineage)
  - [Verify the features in Feature Store](#verify-the-features-in-feature-store)
  - [Exploratory data analysis (EDA)](#exploratory-data-analysis-eda)
  - [Data validation](#data-validation)
- [Model training and inferencing](#model-training-and-inferencing)
  - [Model training](#model-training)

>   - [Model inferencing](#model-inferencing)

Reviewer: Update line 53:

In addition to the main flow, there are optional steps for performing 'exploratory data analysis' and 'data validations' (illustrated by dotted lines in the diagram). These features are not part of the main processing steps, but are covered separately under [Additional Features](#additional-features).

Author: OH! Forgot to update this part accordingly, thanks for the reminder!

2. Create a Fabric lakehouse

   Within the workspace, create a new lakehouse by clicking the 'New' button on the top of the page. Refer to [this documentation](https://learn.microsoft.com/fabric/onelake/create-lakehouse-onelake#create-a-lakehouse) for detailed instructions.

3. Import the sample notebooks

   After the lakehouse is created, go back to the workspace. Click the 'New' button from the **Data Engineering** or the **Data Science** homepage and select 'Import notebook' from the menu. Navigate to `src/notebooks` folder under the repo, and select all the notebooks to import:

   ![import all notebooks](./images/import_all_notebooks.png)

   And below are brief introductions for all notebooks.

> 

Reviewer: I suggest adding links to the notebooks in the table.

Author: Done! :- )

   | Notebook | Description |
   |----------|-------------|

>    | data_ingestion | Ingest data from the `Landing Zone` to the `Staging Zone` |
>    | exploratory_data_analysis | Explore and analyze data in `Staging Zone` to discover patterns, spot anomalies, test a hypothesis, or check assumptions by data scientists according to data analyzing or machine learning model training requirements |
>    | data_validation | Validate data in `Staging Zone` by user-defined rules using [Great Expectations](https://docs.greatexpectations.io/docs/) |
>    | data_cleansing | Cleanse data in `Staging Zone` and sink them to `Standardized Zone` |
>    | data_transformation | Transform cleansed data and sink them to `Standardized Zone` |
>    | feature_set_registration | Create features based on transformed data and register them to Azure ML managed feature store |
>    | data_catalog_and_lineage | Utility functions for registering data assets and lineage to Microsoft Purview |
>    | model_training | Train ML models using features values retrieved from Azure ML managed feature store |
>    | model_inferencing | Do some inferencing based on trained model |
>    | feature_set_retrieval | Utility functions for retrieving features values for model training |
>    | utils | The common utility functions |

Reviewer: ```suggestion
   | data_ingestion | Ingest data from the `Landing Zone` to the `Staging Zone`. |
   | exploratory_data_analysis | Explore and analyze data in `Staging Zone` to discover patterns, spot anomalies, test a hypothesis, or check assumptions by data scientists according to data analyzing or ML model training requirements. |
   | data_validation | Validate data in `Staging Zone` by user-defined rules using [Great Expectations](https://docs.greatexpectations.io/docs/) library. |
   | data_cleansing | Cleanse data in `Staging Zone` and sink it to `Standardized Zone`. |
   | data_transformation | Transform cleansed data and sink it to `Standardized Zone`. |
   | feature_set_registration | Create features based on transformed data and register them to Azure ML managed feature store. |
   | data_catalog_and_lineage | Utility functions for registering data assets and lineage to Microsoft Purview. |
   | model_training | Train ML models using features values retrieved from Azure ML managed feature store. |
   | model_inferencing | Perform inferencing based on trained model. |
   | feature_set_retrieval | Utility functions for retrieving features values for model training. |
   | utils | Common utility functions. |
```

  ![feature_lineage](./images/managed_feature_store.gif)

> ### Exploratory data analysis (EDA)

Reviewer: Few comments:
- Suggest moving it to as an H2 header after "Model training and inferencing" as below:

## Model training and inferencing

## Additional (optional) Features

### Exploratory data analysis (no need to put the short form 'EDA' here)

### Data Validations

> 

Reviewer: Ordering of the content. Suggest, first giving a brief overview of what EDA is and then add the notebook specific detail.

Also, provide link to the notebook at-least once. i.e, 

replace `exploratory_data_analysis` with [exploratory_data_analysis.ipynb](./src/notebooks/exploratory_data_analysis.ipynb)

And do the same for data validation step.

Author: Hi @promisinganuj, 
Thanks for your great suggestions as always! 
Made some updates to address several of them, thanks for your help to check again! :- )

- Configure a [Checkpoint](https://docs.greatexpectations.io/docs/terms/checkpoint) and run the Expectation suite, a checkpoint provides a convenient abstraction for bundling the validation of a batch (or batches) of data provided by one (or several) Batch Requests, against one or several Expectation Suites.
- Report the data quality logs to [Azure Monitor Application Insights](https://learn.microsoft.com/en-us/azure/azure-monitor/app/app-insights-overview)

Before running this notebook, please [find the connection string](https://learn.microsoft.com/en-us/azure/azure-monitor/app/sdk-connection-string?tabs=dotnet5#find-your-connection-string) of your Azure Monitor Application Insights resource and assign it to variable `AZURE_MONITOR_SECRET` in the last cell of the [data_validation](./src/notebooks/data_validation.ipynb) notebook.

```python
# Report Data Quality Metrics to Azure Monitor using python Azure Monitor open-census exporter 
import logging
import time
from opencensus.ext.azure.log_exporter import AzureLogHandler

AZURE_MONITOR_SECRET = "InstrumentationKey=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
```

> For more details, please refer to this article, [Data Validation with Great Expectations on Microsoft Fabric](https://medium.com/@756025472/enhancing-data-validation-with-great-expectations-and-sending-logs-to-azure-monitor-19ba7d345d20).

Reviewer: ```suggestion
For more details, refer to the article, [Data validation with Great Expectations on Microsoft Fabric](https://medium.com/@756025472/enhancing-data-validation-with-great-expectations-and-sending-logs-to-azure-monitor-19ba7d345d20).
```
